src/EntryPoint.sol

```diff
@@ -168,113 +139,101 @@ contract EntryPoint is EIP712, UUPSUpgradeable, Ownable, ReentrancyGuard {
+            if or(lt(shr(6, mul(gas(), 63)), add(g, _INNER_GAS_OVERHEAD)), shr(64, g)) {
+                mstore(0x00, 0x1c26714c) // `InsufficientGas()`.
+                revert(0x1c, 0x04)
+            }
+
+            let m := mload(0x40) // Grab the free memory pointer.
+            // Copy the encoded user op to the memory to be ready to pass to the self call.
+            calldatacopy(add(m, 0x40), encodedUserOp.offset, encodedUserOp.length)
+            let s := add(m, 0x1c) // Start of the calldata in memory to pass to the self call.
+            let n := add(encodedUserOp.length, 0x44) // Length of the calldata to the self call.
+
+            // To prevent griefing, we need to do two non-reverting gas-limited calls.
+            // Even if the verify and call fails, which the gas will be burned, 
+            // the payment has already been made and can't be reverted.
+            for {} 1 {} {
+                // 1. Pay.
+                mstore(m, 0x1a3de5c3) // `_pay()`.
+                mstore(0x00, 0) // Zeroize the return slot.
+                let gCapped := xor(g, mul(xor(g, _PAYMENT_GAS_CAP), lt(_PAYMENT_GAS_CAP, g))) // `min`.
+                if iszero(call(gCapped, address(), 0, s, n, 0x00, 0x20)) {
+                    err := mload(0x00)
+                    if iszero(returndatasize()) { err := shl(224, 0xbff2584f) } // `PaymentError()`.
+                    break
+                }
+                paymentAmount := mload(0x00)
+                let gUsed := sub(gStart, gas())
+                let gLeft := mul(sub(g, gUsed), gt(g, gUsed))
+                // 2. Verify and call.
+                mstore(m, 0xe235a92a) // `_verifyAndCall()`.
+                mstore(0x00, 0) // Zeroize the return slot.
+                if iszero(call(gLeft, address(), 0, s, n, 0x00, 0x20)) {
+                    err := mload(0x00)
+                    if iszero(returndatasize()) { err := shl(224, 0xad4db224) } // `VerifiedCallError()`.
+                    break
+                }
+                break
+            }
+        }
+        uint256 gUsed = Math.rawSub(gStart, gasleft());
+        uint256 paymentPerGas = u.paymentPerGas;
+        if (paymentPerGas == uint256(0)) paymentPerGas = type(uint256).max;
+        uint256 finalPaymentAmount = Math.min(
+            paymentAmount,
+            Math.saturatingMul(paymentPerGas, Math.saturatingAdd(gUsed, _REFUND_GAS))
+        );
+        address paymentRecipient = u.paymentRecipient;
+        if (paymentRecipient == address(0)) paymentRecipient = address(this);
+        if (LibBit.and(finalPaymentAmount != 0, paymentRecipient != address(this))) {
+            TokenTransferLib.safeTransfer(u.paymentToken, paymentRecipient, finalPaymentAmount);   
+        }
+        if (paymentAmount > finalPaymentAmount) {
+            uint256 toRefund = Math.rawSub(paymentAmount, finalPaymentAmount);
+            TokenTransferLib.safeTransfer(u.paymentToken, u.eoa, toRefund);    
+        }
```

Comment:
So instead of doing this, the amount you're `_pay()`-ing is always the `finalPaymentAmount`? `paymentPerGas * gasUsed`.